src/controllers/UploadController.ts

```diff
@@ -0,0 +1,274 @@
+   *     "results": [
+   *       { "cid": "Qm...", "fileName": "success.txt" }
+   *     ],
+   *     "failed": [
+   *       {
+   *         "fileName": "failed.txt",
+   *         "error": "File exceeds size limit"
+   *       }
+   *     ]
+   *   }
+   * }
+   * ```
+   *
+   * No Files Error Response (400):
+   * ```json
+   * {
+   *   "success": false,
+   *   "message": "No files uploaded",
+   *   "uploadStatus": "none",
+   *   "errors": {
+   *     "upload": "No files uploaded"
+   *   }
+   * }
+   * ```
+   *
+   * Upload Failed Error Response (500):
+   * ```json
+   * {
+   *   "success": false,
+   *   "message": "Upload failed",
+   *   "errors": {
+   *     "upload": "Failed to upload file"
+   *   }
+   * }
+   * ```
+   */
+  @Post()
+  @SuccessResponse(201, "Upload successful")
+  public async upload(
+    @UploadedFiles("files") files?: Express.Multer.File[],
+    @FormField() jsonData?: string,
+  ): Promise<UploadResponse> {
+    try {
+      const storage = await StorageService.init();
+
+      if (jsonData) {
+        console.debug("Got JSON data for future use");
+      }
+
+      const blobs = files?.map(
+        (file) => new Blob([file.buffer], { type: file.mimetype }),
+      );
+
+      if (!files || !blobs) {
+        throw new NoFilesUploadedError();
+      }
+
+      const uploadResults = await Promise.allSettled(
+        files.map(async (file, index) => {
+          try {
+            const result = await storage.uploadFile({
+              file: blobs[index],
+            });
+            return {
+              cid: result.cid,
+              fileName: file.originalname,
+            };
+          } catch (error) {
+            throw new SingleUploadFailedError(
+              file.originalname,
+              (error as Error).message,
+            );
+          }
+        }),
+      );
+
+      const successful = uploadResults
+        .filter(isSuccessfulUpload)
+        .map((result) => result.value);
+
+      const failed = uploadResults.filter(isFailedUpload).map((result) => {
+        const error = result.reason as SingleUploadFailedError;
+        return {
+          fileName: error.fileName,
+          error: error.errorDetail,
+        };
+      });
+
+      if (failed.length > 0) {
+        const data = {
+          results: successful,
+          failed: failed,
+        };
+
+        if (failed.length === uploadResults.length) {
+          throw new UploadFailedError("All uploads failed", {
+            results: [],
+            failed: failed, // Preserve all failed upload details
+          });
+        }
+
+        throw new PartialUploadError("Some uploads failed", data);
+      }
+
+      this.setStatus(201);
+      return {
+        success: true,
+        message: "Upload successful",
+        uploadStatus: "all",
+        data: {
+          results: successful,
+          failed: [],
+        },
+      };
+    } catch (error) {
+      if (error instanceof FileUploadError) {
```

Comment:
Non-blocker: there's enough complexity here to warrant a separate class or function to build this return object. Again, not blocking, just a bit of a remark :)

Reply:
Ok, for the next round of cleanups